samples/MQTTSN/src/paho_iot_hub_telemetry_example_qos_0.c

```diff
@@ -0,0 +1,265 @@
+      NULL));
+
+  return AZ_OK;
+}
+
+// This sample implements QoS 0
+int main(int argc, char** argv)
+{
+	int rc = 0;
+	int mysock;
+	unsigned char buf[500];
+	int buflen = sizeof(buf);
+	MQTTSN_topicid topic;
+	MQTTSNString topicstr;
+	int len = 0;
+	int retained = 0;
+	
+	// Default
+	char *host = "127.0.0.1";
+	int port = 10000; // use unicast port if sending a unicast packet
+
+	MQTTSNPacket_connectData options = MQTTSNPacket_connectData_initializer;
+	unsigned short topicid;	
+
+	// Create a unicast UDP socket
+	mysock = transport_open();
+	if(mysock < 0)
+		return mysock;
+
+	// Read for optional destination address and port
+	if (argc > 1)
+		host = argv[1];
+
+	if (argc > 2)
+		port = atoi(argv[2]);
+
+	printf("Connecting to hostname %s port %d\n", host, port);
+
+	// Read in the necessary environment variables and initialize the az_iot_hub_client
+	if (az_failed(rc = read_configuration_and_init_client()))
+	{
+		printf("Failed to read configuration from environment variables, return code %d\n", rc);
+		return rc;
+	}
+
+	options.clientID.cstring = device_id;
+
+	// Get topic name that the IoT Hub is subscribed to
+	if (az_failed(
+		rc = az_iot_hub_client_telemetry_get_publish_topic(
+			&client, NULL, topicname, sizeof(topicname), NULL)))
+	{
+		printf("Failed to get publish topic, return code %d\n", rc);
+		return rc;
+	}
+	
+	// CONNECT to MQTTSN Gateway
+	len = MQTTSNSerialize_connect(buf, buflen, &options);
+
+	if (az_failed(
+		rc = transport_sendPacketBuffer(host, port, buf, len)))
+	{
+		printf("Failed to send Connect packet to the Gateway, return code %d\n", rc);
+		return rc;
+	}
+	
+	// Wait for CONNACK from the MQTTSN Gateway
+	if (MQTTSNPacket_read(buf, buflen, transport_getdata) == MQTTSN_CONNACK)
+	{
+		int connack_rc = -1;
+
+		if (MQTTSNDeserialize_connack(&connack_rc, buf, buflen) != 1 || connack_rc != 0)
+		{
+			printf("Failed to receive Connect ACK packet, return code %d\nExiting...\n", connack_rc);
+			goto exit;
+		}
+		else 
+			printf("CONNACK rc %d\n", connack_rc);
+	}
+	else
+	{
+		printf("Failed to connect to the Gateway\nExiting...\n");
+		goto exit;
+	}
+
+	// REGISTER topic name with the MQTTSN Gateway
+	printf("Registering topic %s\n", topicname);
+	int packetid = 1;
+	topicstr.cstring = topicname;
+	topicstr.lenstring.len = strlen(topicname);
+	len = MQTTSNSerialize_register(buf, buflen, 0, packetid, &topicstr);
+	
+	if (az_failed(
+		rc = transport_sendPacketBuffer(host, port, buf, len)))
+	{
+		printf("Failed to send Register packet to the Gateway, return code %d\n", rc);
+		return rc;
+	}
+
+	// Wait for REGACK from the MQTTSN Gateway
+	if (MQTTSNPacket_read(buf, buflen, transport_getdata) == MQTTSN_REGACK) 	
+	{
+		unsigned short submsgid;
+		unsigned char returncode;
+
+		rc = MQTTSNDeserialize_regack(&topicid, &submsgid, &returncode, buf, buflen);
+		if (returncode != 0)
+		{
+			printf("Failed to receive Register ACK packet, return code %d\nExiting...\n", returncode); // TODO
+			goto exit;

```

Comment:
>goto exit; [](start = 3, length = 10)

Generally we do not use goto. Lets restructure our code then you do not need goto. After you create small function then it will be much cleaner.